### PR TITLE
fix(model-list): read the Ollama context window from /api/show

### PR DESCRIPTION
**Pull request by DeJeune**
> ### Branch strategy
>
> - Active development targets `main`.

### What this PR does

Before this PR:

`ollamaFetcher` built models from `GET /api/tags`, which carries no context length, so every locally discovered Ollama model had `contextWindow` undefined. `buildOllamaProviderOptions` is conditional on it:

```ts
...(model.contextWindow ? { options: { num_ctx: model.contextWindow } } : {})
```

so no `num_ctx` was ever sent and Ollama sized the request by available VRAM — [4k below 24 GiB, 32k up to 48 GiB, 256k above](https://docs.ollama.com/context-length). An Agent session's tool preamble overruns 4k on its own.

`chatPrompt` then strips messages from the front until the prompt fits, re-rendering each time and aborting on the first render error:

```go
for i := 0; i <= lastMsgIdx; i++ {
    p, err := renderPrompt(m, append(system, msgs[i:]...), tools, think)
    if err != nil { return "", nil, err }
    if ctxLen <= opts.NumCtx { currMsgIdx = i; break }
    if i == lastMsgIdx { currMsgIdx = lastMsgIdx; break }   // always keep the last message
}
```

Once the user's turn is stripped, every remaining `user` message is a tool result, and Qwen3.8's renderer rejects that (`model/renderers/qwen35.go`):

```go
for _, message := range messages {
    if message.Role != "user" { continue }
    content := strings.TrimSpace(r.renderContent(message, 0))
    if !(strings.HasPrefix(content, "<tool_response>") && strings.HasSuffix(content, "</tool_response>")) {
        foundUserQuery = true
        break
    }
}
if !foundUserQuery { return fmt.Errorf("no user query found in messages") }
```

which surfaces as `chat prompt error` → `POST /api/chat 500` → an opaque 500 from the gateway. Ollama's own guidance puts agent and coding workloads at 64k+, which only the model's real window can satisfy.

After this PR:

Model listing reads the trained window per model from `POST /api/show` and stores it as `contextWindow`, so `num_ctx` is sent and the request runs at the window the model was trained for. `model_info` keys are architecture prefixed, so the architecture is read first:

```ts
const architecture = modelInfo?.['general.architecture']       // e.g. "qwen3"
const contextLength = modelInfo?.[`${architecture}.context_length`]
```

A model whose `/api/show` call fails or reports a mismatched architecture still lists, just without a window.

Refs #18643

### Why we need it and why it was done in this way

`/api/show` is the only endpoint that reports context length; `/api/tags` does not, which is why the field was empty in the first place. The calls run in parallel per model against a local server and read manifest metadata rather than loading weights.

The following tradeoffs were made:

- One extra request per model at listing time. They are concurrent and local.
- Requesting the model's full trained window allocates a correspondingly larger KV cache. This matches Ollama's own advice ("For best performance, use the maximum context length for a model") and is the behaviour Cherry already had for any model with a configured `contextWindow` — this change only extends it to models that were previously invisible to it.
- Reading the architecture prefix from `general.architecture` means a model that omits it reports no window rather than guessing across `model_info` keys.

The following alternatives were considered:

- **Substituting a fixed `num_ctx` when the window is unknown was rejected, and is worth recording.** It reads as a safe floor but is a regression: with `num_ctx` absent Ollama sizes by VRAM, so any constant small enough to be safe on a laptop (8k) would *shrink* the window on a 48 GiB machine that would otherwise have been given 256k. Omitting the parameter is the better default when the real value cannot be read, so `buildOllamaProviderOptions` stays conditional.
- Parsing the `parameters` block from `/api/show` for a Modelfile `num_ctx` was rejected: that is a per-Modelfile default, not the model's capability, and `contextWindow` means the latter.
- Calling `/api/show` lazily at request time rather than at listing time was rejected because it puts a round trip on the hot path for a value that only changes when the model is pulled.

Links to places where the discussion took place: https://github.com/CherryHQ/cherry-studio/issues/18643

### Breaking changes

None.

### Special notes for your reviewer

The response shape is pinned by [Ollama's OpenAPI spec for `/api/show`](https://docs.ollama.com/api/show) — request body `{ model }`, response `model_info` with `general.architecture` and `<arch>.context_length`.

This is independent of #18780, #18697 and #19042, which fix three other defects behind the same report. It is the first one that addresses the `no user query found in messages` failure, and it is deliberately marked `Refs` rather than `Fixes`: the mechanism is established from Ollama's source, but no live Agent-plus-Ollama reproduction has been run against it, and the interaction is a heuristic in Ollama that can still fail on a long enough session.

Validation completed:

- `vitest run --project main src/main/ai` → 301 files / 4,480 tests passed
- `tsgo --noEmit -p tsconfig.node.json` clean, oxlint 0/0 across `src/main/ai`
- Three new listing tests: the window is read and sent, a failing `/api/show` still lists the model, and a `context_length` that does not match the reported architecture is ignored

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets `main`
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
Local Ollama models now report their real context window, so requests are no longer capped by Ollama's default sizing. This fixes Agent sessions with tools failing against local Ollama models, where the shorter window caused the conversation to be truncated away.
```

### CODE REVIEW

**Review by zhangjiadi225 — APPROVED** [review]

File: src/main/ai/provider/listModels.ts [hunk 1]
```diff
@@ -185,10 +227,15 @@ const ollamaFetcher: ModelFetcher = {
...
+    const models = dedup(response.models, (m) => m.name)
+    const contextWindows = await Promise.all(
+      models.map((m) => fetchOllamaContextWindow(baseUrl, provider, m.name, signal))
+    )
+    return models.map((m, index) =>
       toModel(m.name, provider, {
         ownedBy: 'ollama',
-        capabilities: m.capabilities?.includes('thinking') ? [MODEL_CAPABILITY.REASONING] : []
+        capabilities: m.capabilities?.includes('thinking') ? [MODEL_CAPABILITY.REASONING] : [],
+        ...(contextWindows[index] ? { contextWindow: contextWindows[index] } : {})
```
Comment (kangfenmao) [inline src/main/ai/provider/listModels.ts]: **[A1] The discovered context window is lost before custom Ollama models reach runtime.**

This value survives in the fetched model, but both automatic sync and manual pull additions pass it through `toCreateModelDto()`, which does not copy `contextWindow`. For an Ollama model that does not match a registry preset, `ModelService` therefore stores `context_window = NULL`; later chat/Agent resolution reads that persisted row, and `buildOllamaProviderOptions()` still omits `num_ctx`. Registry-backed models are masked by catalog enrichment, but custom local tags remain unfixed.

Please forward `model.contextWindow` in `toCreateModelDto()` and add a contract test covering a registry-unmatched fetched model through the create DTO (or persistence) boundary.

File: src/main/ai/provider/__tests__/listModels.test.ts [hunk 2]
```diff
@@ -143,6 +152,50 @@ describe('listModels — Ollama capabilities', () => {
       url: 'http://ollama.test:11434/api/tags'
     })
   })
+
+  it('reads the trained context window from /api/show so num_ctx is not left at Ollama default', async () => {
+    // /api/tags carries no context length; without this the model has no contextWindow, no
+    // num_ctx is sent, and Ollama runs the request at 2048 tokens (#18643).
```
[the remark this answers is not in the document]
Comment (DeJeune) [reply src/main/ai/provider/__tests__/listModels.test.ts]: Fixed in `0935fab3` — you're right, and thanks for catching it.

That `2048` was a stale premise I carried in from the existing comment in `buildOllamaProviderOptions`. Ollama has not defaulted to 2048 for a while; with no `num_ctx` it [sizes by available VRAM](https://docs.ollama.com/context-length) — 4k under 24 GiB, 32k to 48 GiB, 256k above. The comment now says that.

The finding actually changed the PR: my first cut also added a fixed `num_ctx` floor for models whose window could not be read, which reads as safe but is a regression — any constant small enough for a laptop would *shrink* the window on a 48 GiB machine that Ollama would otherwise give 256k. `buildOllamaProviderOptions` therefore stays conditional, and omitting the parameter is now documented as deliberate. The reasoning is recorded under "alternatives considered" in the description.

The 4k figure is what still matters for #18643: an agent's tool preamble overruns it on its own, which is what sets off the truncation.

**Comment by DeJeune** [comment]
The first half of this is a real blocker and is fixed in `4fd56974` — thanks, the PR was inert without it.

`toCreateModelDto` builds a whitelist DTO and never carried `contextWindow`, so the window read from `/api/show` survived in memory (the drawer even displayed it) and was dropped on the way to the database. `CreateModelSchema` already accepts the field, so forwarding it is the whole fix, plus two tests: a discovered window reaches the DTO, and a model without one still omits the key rather than writing a null.

The second half — existing rows not refreshed — is real too, but I do not think it belongs in this PR, and I want to flag a hazard before anyone adds it:

`reconcileForProvider` is delete-then-insert and `addModels` filters to ids not already stored, so a row that predates this change only picks up a window if the user removes and re-adds it. A backfill is possible (`bulkUpdate` exists), but **it must fill only where the stored window is empty**. `contextWindow` is user-editable, and a sync-time overwrite would silently revert someone who deliberately lowered the window on a machine that cannot hold the model's trained maximum — exactly the case this PR is trying to give users headroom for. That is a product decision about when a background sync may rewrite user-owned rows, and it needs its own change with its own UI story rather than riding along here.

Worth noting the migration cost is low in the meantime: Ollama models are local and free to remove and re-add from the pull-reconcile drawer, and any newly pulled model gets the window immediately.
